File: jetbrains/src/main/java/com/boundaryml/jetbrains_ext/BamlLanguageServerFactory.java
package com.boundaryml.jetbrains_ext;

import com.intellij.openapi.project.Project;
import com.redhat.devtools.lsp4ij.LanguageServerFactory;
import com.redhat.devtools.lsp4ij.client.LanguageClientImpl;
import com.redhat.devtools.lsp4ij.server.StreamConnectionProvider;
import org.jetbrains.annotations.NotNull;

public class BamlLanguageServerFactory implements LanguageServerFactory {

    @Override
    public @NotNull StreamConnectionProvider createConnectionProvider(@NotNull Project project) {
        return new BamlLanguageServer();
    }
//
//    @Override // If you need to provide client specific features
//    public @NotNull LanguageClientImpl createLanguageClient(@NotNull Project project) {
//        return new BamlLanguageServerFactory(project);
//    }

//    @Override // If you need to expose a custom server API
//    public @NotNull Class<? extends LanguageServer> getServerInterface() {
//        return MyCustomServerAPI.class;
//    }

}

File: jetbrains/src/main/kotlin/com/boundaryml/jetbrains_ext/BamlLanguageServerFactory.kt
package com.boundaryml.jetbrains_ext

import com.intellij.openapi.project.Project
import com.redhat.devtools.lsp4ij.LanguageServerFactory
import com.redhat.devtools.lsp4ij.client.features.LSPClientFeatures
import com.redhat.devtools.lsp4ij.server.StreamConnectionProvider

class BamlLanguageServerFactory : LanguageServerFactory {

    override fun createConnectionProvider(project: Project): StreamConnectionProvider {
        return BamlLanguageServer()
    }

    override fun createClientFeatures(): LSPClientFeatures {
        val features = LSPClientFeatures()
        features.setServerInstaller(BamlLanguageServerInstaller()) // customize language server installer
        return features
    }

//    // If you need to provide client specific features
//    override fun createLanguageClient(project: Project): LanguageClientImpl {
//        return BamlLanguageServerFactory(project)
//    }

//    // If you need to expose a custom server API
//    override fun getServerInterface(): Class<out LanguageServer> {
//        return MyCustomServerAPI::class.java
//    }
}
